odl/solvers/advanced/proximal_operators.py

```diff
@@ -103,14 +107,118 @@ def make_diag(step_size):
     return make_diag
 
 
+def proximal_convexconjugate(proximal):
+    """ Calculate the proximal of the dual using Moreau decomposition
+
+    The Moreau identity states that for any convex function ``F`` with
+    convex conjugate ``F^*``, the proximals satisfy
+
+        prox[a * F^*](x) + a * prox[F / a](x / a) = x
+
+    where ``a`` is a scalar step size. Using this, we find the proximal of the
+    convex conjugate
+
+        prox[a * F^*](x) = x - a * prox[F / a](x / a)
+
+    Note that since ``(F^*)^* = F``, this can be used to get the primal from
+    the dual.
+
+    Parameters
+    ----------
+    proximal : `callable`
+        A factory function that when called with a stepsize returns the
+        proximal operator of ``F``.
+
+    Returns
+    -------
+    prox : `callable`
+        Factory for the proximal operator to be initialized
+    """
+
+    def make_convexconjugate_proximal(step_size):
+        """Create proximal for the dual with a given step_size
+
+        Parameters
+        ----------
+        step_size : positive `float`
+            Step size parameter
+
+        Returns
+        -------
+        proximal : `Operator`
+            The proximal operator of ``prox[a * f^*](x)``
+        """
+        prox_other = step_size * proximal(1.0 / step_size) * (1.0 / step_size)
+        return IdentityOperator(prox_other.domain) - prox_other
+
+    return make_convexconjugate_proximal
+
+
+def proximal_composition(proximal, operator, mu):
+    """ Calculate the proximal of functional a composed with unitary operator
+
+    Given an linear `Operator` ``L`` with the property that for a scalar ``mu``
+
+        L^*(L(x)) = mu * x
+
+    And a convex function ``f``, The following identity holds
+
+        prox[f * L](x) = x + 1/mu L^*(prox[mu * f](Lx) - Lx)
+
+    This factory function implements this functionality.
+
+    There is no simple formula for more general operators.
+
+    Parameters
+    ----------
+    proximal : `callable`
+        A factory function that when called with a stepsize returns the
+        proximal operator of ``F``.
+    operator : `Operator`
+        The operator to compose the functional with
+    mu : `Operator.field` element
+        Scalar such that ``operator.adjoint * operator = mu``
+
+    Returns
+    -------
+    prox : `callable`
+        Factory for the proximal operator to be initialized
+
+    Notes
+    -----
+    The function cannot verify that the identity holds, the user needs to
+    verify this.
+    """
+
+    def make_proximal_composition(step_size):
+        """Create proximal for the dual with a given step_size
+
+        Parameters
+        ----------
+        step_size : positive `float`
+            Step size parameter
+
+        Returns
+        -------
+        proximal : `Operator`
+            The proximal operator of ``prox[step_size * f * L](x)``
+        """
+        Id = IdentityOperator(operator.domain)
+        Ir = IdentityOperator(operator.range)
```

Comment:
Oh, I just realized that |iran| may not be the optimal choice either ...

On 2016-05-21 22:25, Jonas Adler wrote:
> In odl/solvers/advanced/proximal_operators.py
> <https://github.com/odlgroup/odl/pull/401#discussion_r64139796>:
> 
>> +
>> +    def make_proximal_composition(step_size):
>> +        """Create proximal for the dual with a given step_size
>> +
>> +        Parameters
>> +        ----------
>> +        step_size : positive `float`
>> +            Step size parameter
>> +
>> +        Returns
>> +        -------
>> +        proximal : `Operator`
>> +            The proximal operator of ``prox[step_size * f * L](x)``
>> +        """
>> +        Id = IdentityOperator(operator.domain)
>> +        Ir = IdentityOperator(operator.range)
> 
> Agree, but these are "obvious" in this 4 line function. Mainly exist to
> make later lines easier.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/odlgroup/odl/pull/401/files/5a721d1d1325be17940d8eeb33e35bdd4879ad8a#r64139796>
> 

-- 

Dr Holger Kohr
KTH - Royal Institute of Technology
Dpt of Mathematics
100 44 Stockholm
Tel +46 8 790-6652
Mobil +46 72 0119332